kv_provider_protostellar.go

```diff
@@ -0,0 +1,337 @@
+	request := &kv_v1.ExistsRequest{
+		Key: string(opm.DocumentID()),
+
+		CollectionName: opm.CollectionName(),
+		ScopeName:      opm.ScopeName(),
+		BucketName:     opm.BucketName(),
+	}
+
+	res, err := p.client.Exists(opm.ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	resOut := ExistsResult{
+		Result: Result{
+			Cas(res.Cas),
+		},
+		docExists: res.Result,
+	}
+
+	return &resOut, nil
+}
+
+func (p *kvProviderProtoStellar) Delete(opm *kvOpManager) (*MutationResult, error) {
+
+	cas := opm.Cas()
+
+	request := &kv_v1.RemoveRequest{
+		BucketName:      opm.BucketName(),
+		ScopeName:       opm.ScopeName(),
+		CollectionName:  opm.CollectionName(),
+		Key:             string(opm.DocumentID()),
+		Cas:             (*uint64)(&cas),
+		DurabilityLevel: memdDurToPs(opm.DurabilityLevel()),
+	}
+
+	res, err := p.client.Remove(opm.ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	mt := psMutToGoCbMut(*res.MutationToken)
+	outCas := res.Cas
+
+	mutOut := MutationResult{
+		mt: &mt,
+		Result: Result{
+			cas: Cas(outCas),
+		},
+	}
+
+	return &mutOut, nil
+}
+
+func (p *kvProviderProtoStellar) Unlock(opm *kvOpManager) error {
+
+	cas := opm.Cas()
+
+	request := &kv_v1.UnlockRequest{
+		BucketName:     opm.BucketName(),
+		ScopeName:      opm.ScopeName(),
+		CollectionName: opm.CollectionName(),
+		Key:            string(opm.DocumentID()),
+		Cas:            (uint64)(cas),
+	}
+
+	_, err := p.client.Unlock(opm.ctx, request)
+
+	return err
+
+}
+
+func (p *kvProviderProtoStellar) Touch(opm kvOpManager) (*MutationResult, error) {
+
+	request := &kv_v1.TouchRequest{
+		BucketName:     opm.BucketName(),
+		ScopeName:      opm.ScopeName(),
+		CollectionName: opm.CollectionName(),
+		Key:            string(opm.DocumentID()),
+		//Expiry:         opm.Expiry(), // TODO: expiry
+	}
+
+	res, err := p.client.Touch(opm.ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	mt := psMutToGoCbMut(*res.MutationToken)
+	outCas := res.Cas
+
+	mutOut := MutationResult{
+		mt: &mt,
+		Result: Result{
+			cas: Cas(outCas),
+		},
+	}
+
+	return &mutOut, nil
+
+}
+
+func (p *kvProviderProtoStellar) GetReplica(opm *kvOpManager) (*GetReplicaResult, error) {
+	request := &kv_v1.GetReplicaRequest{
+		BucketName:     opm.BucketName(),
+		ScopeName:      opm.ScopeName(),
+		CollectionName: opm.CollectionName(),
+		Key:            string(opm.DocumentID()),
+		ReplicaIndex:   uint32(opm.ReplicaIndex()),
+	}
+
+	res, err := p.client.GetReplica(opm.ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	outCas := res.Cas
+
+	docOut := &GetReplicaResult{}
+	docOut.cas = Cas(outCas)
+	docOut.transcoder = opm.Transcoder()
+	docOut.contents = res.Content
+	docOut.flags = uint32(res.ContentType)
+	docOut.isReplica = true
+
+	return docOut, nil
+
+}
+
+// converts memdDurability level to protostellar durability level
+func memdDurToPs(dur memd.DurabilityLevel) *kv_v1.DurabilityLevel {
+	// memd.Durability starts at 1.
```

Comment:
This might be better using a switch state with each level as a case. Just for extra clarity.